### ISSUE TITLE
[R-Forge #1874] Add error message for unsupported column types such as POSIXlt.

### ISSUE THREAD

**Issue by arunsrinivasan** [Rdatatable/data.table#646]
Submitted by: Matt Dowle; Assigned to: Nobody; [R-Forge link](https://r-forge.r-project.org/tracker/index.php?func=detail&aid=1874&group_id=240&atid=978)

In data.table(), as.data.table(), and := when creating new columns.
As reported here :
http://r.789695.n4.nabble.com/problems-with-adding-a-DateTime-column-tp4456838p4456838.html

**Comment by tdhock** [comment]
I confirm that this is still an issue on current master:

```r
date.str <- "1984-03-17"
buggy.dt <- data.table(date.str)

## BUG: currently this warning is confusing (Supplied 11 items to be
## assigned to 2 items of column 'date.lt'). Instead I propose
## converting to a POSIXct object with a warning, as with
## data.table():
buggy.dt[, date.lt := strptime(date.str, "%Y-%m-%d")]
## Warning message:
## In data.table(date = strptime("2015-10-10", "%Y-%m-%d")) :
##   POSIXlt column type detected and converted to POSIXct. We do not recommend use of POSIXlt at all because it uses 40 bytes to store one date.
date.lt <- strptime(date.str, "%Y-%m-%d")

## I expected this should work as in base R:
(month.str <- strftime(date.lt, "%Y-%d-01"))
buggy.dt[, month.str := strftime(date.lt, "%Y-%d-01")]
## Error in as.POSIXlt.default(x, tz = tz) : 
##   do not know how to convert 'x' to class “POSIXlt”

## I also expected this conversion to work as in base R:
(date.ct <- as.POSIXct(date.lt))
buggy.dt[, date.ct := as.POSIXct(date.lt)]
## Error in as.POSIXct.default(date.lt) : 
##   do not know how to convert 'date.lt' to class “POSIXct”
```

Seems like the fix should be easy, just add a special case inside of `[.data.table`:
```r
if(class of RHS of := is POSIXlt) LHS <- as.POSIXct(LHS)
```
However, after looking at the source code I am not sure where would be the most appropriate place to add this. If you can point me to where to add that, I could submit a PR.

**Comment by tdhock** [comment]
btw

```r

## The workaround is to use explicitly use as.POSIXct, but it would be
## nice if data.table did this for us automatically.
buggy.dt[, date.ct := as.POSIXct(strptime(date.str, "%Y-%m-%d"))]
buggy.dt[, month.str := strftime(date.ct, "%Y-%d-01")]
```

**Comment by jangorecki** [comment]
Recycling warning is on point, POSIXlt is type of list, so potentially it could be consumed as such:
```r
library(data.table)
x = as.POSIXlt(Sys.time())
typeof(x)
#[1] "list"
data.table(attr = names(unlist(x)))[, value := x][]
#      attr    value
# 1:    sec 31.91862
# 2:    min       27
# 3:   hour       15
# 4:   mday       27
# 5:    mon       10
# 6:   year      115
# 7:   wday        5
# 8:   yday      330
# 9:  isdst        0
# 10:   zone      GMT
# 11: gmtoff        0
```
Having a wrapper on `strftime` seems to be the best way to deal with it. You can always replace the original `strftime` function to not change the code at all.
```r
strftime = function(...) as.POSIXct(base::strftime(...))
```

**Comment by arunsrinivasan** [comment]
@tdhock thanks. I'm working on making `[.data.table` simpler to access for users (under the hopes that people would be able to provide more patches :-), and it'd be easier for us to maintain things in the future). 

I'll be able to provide more details on where you should look at to provide a PR once that's done. But this definitely needs to be fixed for the next stable release. Agreed.

**Comment by mattdowle** [comment]
Minor fix 85bacc793d33c628841f5fc8fd8f8992092aa80e for Solaris, #1934.
